Fix dim lengths created from coords or ConstantData #5751

LukeLB opened the pull request:
Closes #5751. Fixes dimensions created by add_coord by making dimension length a TensorConstant. I have also added a test for this using some code from #5181. 

I'm setting as WIP as @canyon289 suggested this PR is required for other work to become unblocked and I'm having problems with some failing tests (which could just be due to me running things locally with an incorrect set-up?). So it's worth seeing if they fail on here and then look at fixes, which will be faster than me trying to figure things out on my own!

Depending on what your PR does, here are a few things you might want to address in the description:
+ [x] what are the (breaking) changes that this PR makes?
+ [x] important background, or details about the implementation
+ [x] are the changes—especially new features—covered by tests and docstrings?
+ [x] [consider adding/updating relevant example notebooks](https://github.com/pymc-devs/pymc-examples)
+ [ ] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

Commit:
Require `pm.MutableData` or `add_coord(mutable=True)` for mutable dims

This increases the safety of working with resizable dimensions.
Dims created via `pm.Model(coords=...)` coordinates are now immutable.
Only dimension lengths that are created anew from `pm.MutableData`,
`pm.Data(mutable=True)`, or the underlying `add_coord(mutable=True)`
will become shared variables.

Co-authored-by: Michael Osthege <michael.osthege@outlook.com>

canyon289 commented:
> it's worth seeing if they fail on here and then look at fixes, which will be faster than me trying to figure things out on my own

Great strategy. Thanks for taking such a principled approach to development

It seems that there type being changed somewhere. Might be as simple as updating this test.
https://github.com/pymc-devs/pymc/runs/6435716258?check_suite_focus=true#step:7:80

canyon289 commented on pymc/tests/test_model.py:
```diff
@@ -702,6 +702,13 @@ def test_nested_model_coords():
 
 
+def test_add_coord_fixed_dims():
+    with pm.Model(coords={"feature": [1], "group": ["A", "B"]}):
```
Add a comment to this one since its not obvious from reading how fixed dims are being tested

LukeLB commented:
Failing test looks to be down to sampling error? 
```
test_sample_posterior_predictive_after_set_data_with_coord
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=0.1
E       
E       Mismatched elements: 2 / 2 (100%)
E       Max absolute difference: 0.13669218
E       Max relative difference: 0.02331315
E        x: array([5, 6])
E        y: array([4.89599 , 5.863308]) 
```
And it's highlighted in an outstanding issue #5739.

michaelosthege commented on pymc/model.py:
```diff
@@ -1109,8 +1113,11 @@ def add_coord(
+            if fixed:
+                self._dim_lengths[name] = length or aesara.tensor.constant(len(values))
+            else:
+                self._dim_lengths[name] = length or aesara.shared(len(values))
```
We should add a test to cover this

michaelosthege commented on pymc/tests/test_model.py:
```diff
@@ -702,6 +702,16 @@ def test_nested_model_coords():
+    with pm.Model(coords={"feature": [1], "group": ["A", "B"]}):
+        x = pm.Normal("x", 0, 1, size=(1), dims="feature")
+        y = pm.Normal("y", x[..., None], 1, size=(1, 2), dims=("feature", "group"))
+    assert y.eval().shape == (1, 2)
```
```suggestion
        x = pm.Normal("x", 0, 1, dims="feature")
        y = pm.Normal("y", x[..., None], 1, dims=("feature", "group"))
    assert x.eval().shape == (1,)
    assert y.eval().shape == (1, 2)
```

LukeLB commented:
This test is failing because y is not the correct shape, `y.eval().shape == (1, 1)`. Both feature and group are TensorConstants so the addition to add_coord is doing as intented with respect to type changes. I'm not sure I understand why this is still happening.....
```{python}
def test_add_coord_fixed_dims():
    """
    test that dims does not cause resizing of y. See #5181.
    """
    with pm.Model(coords={"feature": [1], "group": ["A", "B"]}):
        x = pm.Normal("x", 0, 1, dims="feature")
        y = pm.Normal("y", x[..., None], 1, dims=("feature", "group"))
    assert x.eval().shape == (1,)
    assert y.eval().shape == (1, 2)
```

michaelosthege commented:
Maybe do more asserts?

```python
with pm.Model(coords={"feature": [1], "group": ["A", "B"]}):
    assert pmodel.dim_lengths["feature"].eval() == 1
    assert pmodel.dim_lengths["group"].eval() == 2
```

LukeLB commented:
My thinking for that test was to make sure that size and dims can be passed at the same time and we get the expected shape out which will not have been possible before this type change in `add_coord()`. So maybe its worth having two tests, the one you've suggested, and the other we can use the case you highlight https://github.com/pymc-devs/pymc/issues/5181#issuecomment-970573037 to test that size and dims can be passed at the same time.  What do you think?

michaelosthege commented:
When `size/shape` and `dims` are passed at the same time, the `size/shape` takes priority and `dims` are just tracked for labeling the dimensions in the final `xarray` objects.

That's why I suggested to _only_ pass dims, and actually it's a little concerning that the test doesn't pass.

Maybe call `add_coord` directly while passing `fixed`. For the variables depending on it there should be no difference between `aesara.shared()` or `at.constant()` dim lengths.

```python
with pm.Model() as pmodel:
    pmodel.add_coord("feature", [1], fixed=True/False)
    pmodel.add_coord("group", ["A", "B"], fixed=True/False)
    assert pmodel.dim_lengths["feature"].eval() == 1
    assert pmodel.dim_lengths["group"].eval() == 2
    x = pm.Normal("x", 0, 1, dims="feature")
    y = pm.Normal("y", x[..., None], 1, dims=("feature", "group"))
assert x.eval().shape == (1,)
assert y.eval().shape == (1, 2)
```

LukeLB commented:
When I try that all cases of fixed = True/False lead to y.eval().shape == (1, 1).

michaelosthege commented:
Oh, now I see why. The `x[..., None]` creates the RV as `(1, 1)` and because the dimensionality is as expected, there's no further resizing.

This works 👇 because then the RV is `(1,)` and repeated for the "group" dimension.

```python
    # ...
    x = pm.Normal("x", 0, 1, dims="feature")
    y = pm.Normal("y", x, 1, dims=("group", "feature"))
assert x.eval().shape == (1,)
assert y.eval().shape == (2, 1)
```

LukeLB commented:
I think I'm still misunderstanding something here. Wouldn't passing dims like that still be allowed prior to this PR? So I don't understand what it is testing with respect to this PR.

michaelosthege commented:
Yes, that's already allowed, but where this test comes in is in combination with the changes from #5761.

Actually, it's a little unfortunate that these changes are on separate branches.
Your test case should (latest after adding a `pm.set_data`) close the coverage gaps from that other PR.

I'm inclined to merge #5761 even with those coverage gaps (I'll review it really carefully right now).
After that you could just `git rebase origin/main`, add two lines doing `pm.set_data` and assert the `ShapeError` or `ShapeWarning` and then we should be good to go 🤔

michaelosthege commented:
@LukeLB the other PR was merged, so now you should be able to rebase and bring the coverage back up :)

LukeLB commented:
@michaelosthege OK I think I understand. I'll rebase and commit your suggested changes.

Commit:
Remove ShapeWarning in favor of raising ValueError directly

michaelosthege commented on pymc/tests/test_model.py:
```diff
@@ -702,6 +703,61 @@ def test_nested_model_coords():
+        a = pm.Normal("a", mu=[1, 2, 3], dims="fixed")
+        m = pm.MutableData("m", [[1, 2, 3]], dims=("one", "fixed"))
+
+    with pytest.raises(ShapeError):
```
```suggestion
    with pytest.raises(ShapeError, match="was initialized from 'a'"):
```

michaelosthege replied:
Included from the test below, to reduce test runtime.
```suggestion
        m = pm.MutableData("m", [[1, 2, 3]], dims=("one", "fixed"))
        
        # This is fine because the "fixed" dim is not resized
        pm.set_data({"m": [[1, 2, 3], [3, 4, 5]]})

    with pytest.raises(ShapeError):
```

michaelosthege commented on pymc/tests/test_model.py:
```diff
@@ -702,6 +703,61 @@ def test_nested_model_coords():
+        # This is fine because the "fixed" dim is not resized
+        pm.set_data({"m": [[1, 2, 3], [3, 4, 5]]})
+
+
```
```suggestion
```
Let's include those two lines in the test above

michaelosthege commented on pymc/tests/test_model.py:
```diff
@@ -702,6 +703,61 @@ def test_nested_model_coords():
+        x = pm.Normal("x", 0, 1, dims="feature")
+        y = pm.Normal("y", x, 1, dims=("group", "feature"))
+    assert x.eval().shape == (1,)
+    assert y.eval().shape == (2, 1)
```
This is the test case that can be used to assert the `ShapeWarning`:

```python
with pytest.warns(ShapeWarning, match="was initialized from coords"):
    ...
```

LukeLB replied:
This code doesn't raise a ShapeWarning. Should it?

LukeLB replied:
Ah no worries, figured it out.

LukeLB replied:
@michaelosthege I'm struggling to isolate the ShapeWarning, the code 
```python
with pm.Model() as pmodel:
    pmodel.add_coord("shared", [1,2,3], fixed=False)
    m = pm.MutableData("m", [1,2,3], dims=("shared"))

    pm.set_data({"m": [1,2,3,4]}) 
```
raises both ShapeWarning and a ValueError
```
ShapeWarning: You're changing the shape of a variable in the 'shared' dimension which was 
initialized from coords. Make sure to update the corresponding coords, otherwise you'll get shape issues.
  warnings.warn(

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/var/folders/yw/hvmlf6sd7yv3040wn8str6540000gn/T/ipykernel_1859/831241587.py in <module>
      4     m = pm.MutableData("m", [1,2,3], dims=("shared"))
      5 
----> 6     pm.set_data({"m": [1,2,3,4]})

pymc/pymc/model.py in set_data(new_data, model, coords)
   1798 
   1799     for variable_name, new_value in new_data.items():
-> 1800         model.set_data(variable_name, new_value, coords=coords)
   1801 
   1802 

pymc/pymc/model.py in set_data(self, name, values, coords)
   1220                 if original_coords is not None:
   1221                     if new_coords is None:
-> 1222                         raise ValueError(
   1223                             f"The '{name}' variable already had {len(original_coords)} coord values defined for "
   1224                             f"its {dname} dimension. With the new values this dimension changes to length "

ValueError: The 'm' variable already had 3 coord values defined for its shared dimension. 
With the new values this dimension changes to length 4, so new coord values for the shared dimension are required.
```
I also tried the example from #5760,
```python
x = np.random.normal([13,14,12],[2,1,3],size=(100,3))
x_new = x = np.random.normal([14,15,11],[2,1,3],size=(100,3))

with pm.Model(coords=coords) as model:
    model.add_coord("category", ["A","B","C"], fixed = False)
    model.add_coord("data_point", list(range(1,101)), fixed = False)
    x = pm.MutableData("x",x,dims=["data_point","country"])
    mu = pm.Normal("mu",mu=[12,12,12],sigma=2)
    sigma = pm.Exponential("sigma",lam=[1,1,1])
    obs = pm.Normal("obs",mu=mu,sigma=sigma,observed=x)
with model:
    pm.set_data({"x":x_new})
```
Which doesn't raise any errors or warnings. Not sure how to proceed here...

michaelosthege replied:
The second example doesn't resize the `"data_point"` dimension, that's why it works without problems.

The traceback you posted above looks like `set_data` first warns, but then proceeds to raise a value error anyhow.

Obviously I overlooked this code path in #5761.
We should try to answer the following question:

__Is there a valid scenario where someone might want to resize a dimension that was initially created from `coords`, but can not provide new coordinate values in the same call to `set_data`?__

I feel like with the `ShapeWarning` we assume that there is such a scenario, so we should come up with a corresponding test.
But if in fact this scenario does not exist, we can just raise the `ValueError` directly.

LukeLB replied:
@michaelosthege I would say we should raise the ValueError directly for this scenario. I can't imagine a situation wherein someone would both be using `add_coord()` directly and then not be able to use the coord they had previously passed to `add_coord()` in `set_data()`.

michaelosthege replied:
Agreed. And let's make sure to have `match="..."` for all the error/warning asserts :)

michaelosthege reviewed (commented):
Can you add test cases for the `ShapeWarning` and `ValueError`?

With `match=""` one can make sure that the exception is the correct one. That's important because there are two different `ShapeError`s that one could run into..

We should aim to cover all code paths from the diff here https://github.com/pymc-devs/pymc/pull/5761/files

michaelosthege commented on pymc/tests/test_model.py:
```diff
@@ -702,6 +703,78 @@ def test_nested_model_coords():
+            # correctly passes coords during resize and so warns user with ShapeWarning
+            pm.set_data({"m": [1, 2, 3, 4]}, coords={"shared": [1, 2, 3, 4]})
+
+
```
The warning is now obsolete. In this test you were already passing new `coords` which the warning recommended to do..
```suggestion
```

LukeLB replied:
My bad I should of read that Warning more carefully, I thought it was warning that there was a change in shape!

michaelosthege commented on pymc/tests/test_model.py:
```diff
@@ -702,6 +703,78 @@ def test_nested_model_coords():
+        pmodel.add_coord("shared", [1, 2, 3], fixed=False)
+        m = pm.MutableData("m", [1, 2, 3], dims=("shared"))
+    with pytest.raises(ValueError, match="'m' variable already had 3"):
+        with pmodel:
```
No need to close & open the modelcontext
```suggestion
        with pytest.raises(ValueError, match="'m' variable already had 3"):
```

michaelosthege reviewed (changes requested):
I can't make the change suggestion, but the whole block containing the `ShapeWarning` can be removed..

michaelosthege reviewed (approved):
Thanks @LukeLB !

This brings the coverage back up should lead to fewer shape problems long term.

michaelosthege commented on pymc/tests/test_model.py:
```diff
@@ -702,6 +703,64 @@ def test_nested_model_coords():
+        x = pm.Normal("x", 0, 1, dims="feature")
+        y = pm.Normal("y", x, 1, dims=("group", "feature"))
+    assert x.eval().shape == (1,)
+    assert y.eval().shape == (2, 1)
```
```suggestion
def test_coords_and_constantdata_create_fixed_dims():
    """
    When created from `pm.Model(coords=...)` or `pm.ConstantData`
    a dimension should be fixed in length.
    """
    with pm.Model(coords={"group": ["A", "B"]}) as m:
        x = pm.ConstantData("x", [0], dims="feature")
        y = pm.Normal("y", x, 1, dims=("group", "feature"))
    assert isinstance(m._dim_lengths["feature"], TensorConstant)
    assert isinstance(m._dim_lengths["group"], TensorConstant)
    assert x.eval().shape == (1,)
    assert y.eval().shape == (2, 1)
```

michaelosthege reviewed (changes requested):
ah damn, I just realized that we forgot to actually _use_ the new `fixed` kwarg.

I commented a change in one of the tests that should currently fail.
Let me know if you want me to take over - you put in a lot of effort already and I don't want to get demotivated by all this back and forth

LukeLB commented:
> ah damn, I just realized that we forgot to actually _use_ the new `fixed` kwarg.
> 
> I commented a change in one of the tests that should currently fail. Let me know if you want me to take over - you put in a lot of effort already and I don't want to get demotivated by all this back and forth

Not at all! The longer it goes the more I learn :) Thanks for the offer though.

LukeLB commented:
Why does the assert statement on line 758 fail? The `ConstantData()` function suggests that it returns a `TensorConstant`.

michaelosthege commented:
> Why does the assert statement on line 759 fail? Its created from coord so should be a `TensorConstant` but is `Subtensor{int64}.0`.

The `Subtensor{int64}.0` is because from `pm.ConstantData([1,2,3], dims="foo")` it's effectively doing this: `dim_lengths["foo"] = at.constant([1,2,3]).shape[0]`.

So what needs to be done is a change in the logic behind `pm.ConstantData` (or rather `pm.Data`) to determine the constant data's shape non-symbolically first, and then call `add_coord(..., fixed=True)`.


And I should add that `pm.MutableData([1,2,3], dims="bla")` becomes: `dim_lengths["bla"] = aesara.shared([1,2,3]).shape[0]`.

canyon289 commented on pymc/model.py:
```diff
@@ -1084,9 +1085,12 @@ def add_coord(
             Coordinate values or ``None`` (for auto-numbering).
             If ``None`` is passed, a ``length`` must be specified.
+        fixed : bool
+            Whether dimensions created from coords should be fixed.
```
FYI Fixed has could have a double meaning here, Fixed meaning immutable, or fixed meaning "fixed from broken. Suggest expanding the docstring. @michaelosthege let me know if this makes sense

```suggestion
            Whether dimensions created from coords should be fixed to an immutable value.
```

canyon289 commented on pymc/tests/test_model.py:
```diff
@@ -702,6 +703,67 @@ def test_nested_model_coords():
+    """
+    with #5751 dimensions set by add_coord() are default TensorConstant type
+    and so have fixed dimensions. Trying to resize them will raise a ShapeError,
+    which this test checks for.
```
```suggestion
    which this what test checks for.
```

canyon289 commented on pymc/tests/test_model.py:
```diff
@@ -702,6 +703,67 @@ def test_nested_model_coords():
 
+def test_shape_error_from_fixed_dims():
+    """
+    with #5751 dimensions set by add_coord() are default TensorConstant type
```
Nit: Change docstrings to match numpy style. Notably first line contains summary, and subsequent notes fit in a paragraph below.

https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard

LukeLB commented:
OK so I think I have a solution which I have pushed but I'm running into a wall with a failing test which I can't seem to fix. Unfortunatley I won't be able to work on this till after the weekend so if there is a time pressure on this then it will be worth someone else picking it up.

michaelosthege commented:
Thanks @LukeLB, I will try to pick it up in the evening

canyon289 commented:
@LukeLB this weekend is fine! You've made great progress on this and don't want to take away joy of getting it merged!

michaelosthege committed:
Optionally pass `coords` when creating `pm.Data` variables

Needed to create resizable dims with coordinate values,
because coords passed to the model will become immutable dims.
See https://github.com/pymc-devs/pymc/pull/5763.

Also, because `pm.Data` variables can be N-dimensional,
the coordinate values can't reliably be taken from the value.

michaelosthege commented:
I rebased the branch and squashed the history into three logically separate commits:
1. ✔ Removing the ShapeWarning in favor of raising earlier
2. ✔ Adding optional `coords` to the `pm.Data` API, because we were taking coordinate values from the data variable, which does not work for >1 dimensions.
3. ✔ Addition of the `fixed` kwarg (the actual goal of this PR)

I will force-push them one by one so we can confirm that the first two commits don't create new problems.

michaelosthege commented:
@LukeLB @canyon289 @ricardoV94 please review and note the new changes I made in the `pm.Data` API

ricardoV94 commented on pymc/data.py:
```diff
@@ -499,17 +505,22 @@ def determine_coords(model, value, dims: Optional[Sequence[str]] = None) -> Dict
         for size, dim in zip(value.shape, dims):
             coord = model.coords.get(dim, None)
-            if coord is None:
+            if coord is None and dim is not None:
```
I don't have the full context of this function but in case it's appropriate, can we also take advantage of static shape of Aesara\`s variables? Every entry in `tensorvariable.type.shape` should either be `None` or an integer for the size of that dimension.

michaelosthege replied:
The context of this function is to extract coords from data variales such as `pd.Series` or `pd.DataFrame` that have an index.
I extended the signature to also extract `dims`, because indexes can be named.

It would also be neat to consider `xarray` variables here, ebcause they are a 1:1 match for `pm.Data`.

The `and dim is not None` I added is because `dims` elements can be `None` and then the `coords[None]` doesn't make sense.

What you're talking about with the static shapes might be more relevant in the `pm.Data` function itself, where I created the local `xshape` variable.

ricardoV94 commented on pymc/data.py:
```diff
@@ -665,15 +685,28 @@ def Data(
+        if not mutable:
+            # Use the dimension lengths from the before it was tensorified.
+            # These can still be tensors, but in many cases they are numeric.
+            xshape = np.shape(arr)
```
doesn't matter much but `x.type.shape` should have the same info

ricardoV94 commented on pymc/data.py:
```diff
@@ -665,15 +685,28 @@ def Data(
+                    # Note: Coordinate values can't be taken from
+                    # the value, because it could be N-dimensional.
+                    values=coords.get(dname, None),
+                    fixed=not mutable,
```
can't we use the same keyword so that we don't need to negate booleans?

canyon289 replied:
Id also prefer the same keyword throughout the codebase so its clear how userapi flows through the rest of the codebase

michaelosthege replied:
Ja, you're right, of course.

Done.

ricardoV94 reviewed (approved):
Looks good to me, left a small question and a suggestion about using the same keyword internally (fixed vs mutable)

michaelosthege reviewed (approved):
I feel 👍 about merging as long as the tests go ✔.

canyon289 commented:
Me too!

canyon289 commented:
Congrats @LukeLB! Thank you for leading this PR

LukeLB commented:
@canyon289 @michaelosthege @ricardoV94 thanks all!! It's been a pleasure :)
